File: Cplusplus/0046_permutations/0046_permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        if (nums.size() == 0) 
            return res;
        sort(nums.begin(), nums.end());
        permutation(res, nums, 0);
        return res;
    }
    
    void permutation(vector<vector<int>> &res, vector<int> &nums, int pos) {
        if (pos == nums.size()) {
            vector<int> temp;
            for (int n : nums)
                temp.push_back(n);
            res.push_back(temp);
        }
        else {
            for (int i = pos; i < nums.size(); i++) {
                swap(nums[i], nums[pos]);
                permutation(res, nums, pos + 1);
                swap(nums[i], nums[pos]);
            }
        }
    }
};

File: Java/0046_permutations/0046_permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if (nums.length == 0) 
            return res;
        Arrays.sort(nums);
        permutation(res, nums, 0);
        return res;
    }
    
    private void permutation(List<List<Integer>> res, int[] nums, int pos) {
        if (pos == nums.length) {
            List<Integer> tempList = new ArrayList<Integer>();
            for (int n : nums)
                tempList.add(n);
            res.add(tempList);
        }
        else {
            for (int i = pos; i < nums.length; i++) {
                swap(i, pos, nums);
                permutation(res, nums, pos + 1);
                swap(i, pos, nums);
            }
        }
    }
    
    private void swap(int i, int j, int[] nums) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}
